### ISSUE TITLE
"onSuccess" script runs before dts compile done.

### ISSUE THREAD

**Issue by Misaka-0x447f** [egoist/tsup#700]
![image](https://user-images.githubusercontent.com/15797507/186830902-d7fa6a8b-bac6-4c36-8be1-1ca21fc53c07.png)

**Comment by egoist** [comment]
yeah this was initially added to let you run the generated js file, so we didn't wait for the dts build to succeed to call this function.

can you explain why you need this? maybe we can call this function again when the dts build is ready:

```js
onSuccess: ({ isDtsReady }) => {
  // isDtsReady: boolean
}
```

**Comment by Misaka-0x447f** [comment]
> yeah this was initially added to let you run the generated js file, so we didn't wait for the dts build to succeed to call this function.
> 
> can you explain why you need this? maybe we can call this function again when the dts build is ready:
> 
> ```js
> onSuccess: ({ isDtsReady }) => {
>   // isDtsReady: boolean
> }
> ```

I need let the developer knows compile success and is now waiting for change, but not done without a process exit. Maybe add a hardcoded prompt when everything’s done  in watch mode is good enough.

**Comment by alii** [comment]
> can you explain why you need this? maybe we can call this function again when the dts build is ready:

@egoist reasons for needing this functionality in #856

**Comment by RichiCoder1** [comment]
^ basically the same reason. Need to rename the types (https://twitter.com/robpalmer2/status/1634702219025981440?s=20). Currently doing `tsup && script-that-renames.js`, but it'd be really nice to consolidate that. (And/or have it be something tsup does)

**Comment by victor9000** [comment]
I'm running into this problem as well while using onSuccess to trigger other tasks. Currently the event is being called twice for cjs & dts builds, where I need it to be called once after everything has completed.

**Comment by victor9000** [comment]
Also, I don't think using `isDtsReady` would address my problem.  I ran into cases where dts completed before cjs, so I would still end up in the same place.

**Comment by coopbri** [comment]
What I ended up doing was running the build without DTS generation, then running a DTS-only build in `onSuccess`. This solution is also a bit of a hack and loses some concurrency, but it works. From there, you can tap into the lifecycle however you want. Implemented here: https://github.com/animareflection/ui/blob/bd10752060d082e51c2d489974d247f6b3d36944/tsup.config.ts#L46-L53

**Comment by brc-dd** [comment]
One can also do something like this:

```ts
// tsup.config.ts

import { defineConfig } from 'tsup'

export default defineConfig({
  ...
})

process.on('beforeExit', (code) => {
  if (code === 0) {
    // do something here...
  }
})
```

**Comment by NickBolles** [comment]
We're using onsuccess to call `yalc publish` for a dev command. We see the JS code get updated in the consuming app, but not the typings until the next build. calling onSuccess again when dts is complete, or having another handler for dts completion would be helpful for this usecase too.

**Comment by its-lee** [comment]
I'm using this to determine when it's safe for downstream code to locally npm install a package as part of my workflow. As things stand, it only pulls in the non .d.ts code which doesn't really work for Typescript consumers.

**Comment by nerdyman** [comment]
I ran into this issue using an ESBuild plugin that generates custom `.js` and `d.ts` files at the end of the build. It seems that when `clean` is enabled it doesn't clear out the `d.ts` files until the dts stage runs so I had a really weird issue where the `.js` files were being written but the generated `.d.ts` files never appeared because the DTS stage was deleting them after the `onSuccess` fired.

Would be great to have an option to wait for the DTS stage to finish too.

**Comment by zackdotcomputer** [comment]
I ran into a variant of this issue with the [suggested workaround for .d.ts.map files](https://github.com/egoist/tsup/issues/564) from this repo, where the command is also not waiting until `onSuccess` returns. That means that, because the `.d.ts` files are created by `onSuccess`, a build cascade in a monorepo management tool like `turbo` winds up dying because the next layer of the build starts building before the type definitions are finished.

**Comment by someengineer-nordsec** [comment]
Its 2026 and we have `tsdown` now, but if anyone is looking for cheap trick

```
"dev": "node watch-tsup.js",
"yalc-release": "yalc push --force=true",
```


```js
/* eslint-disable no-undef */
/* eslint-disable no-console */
const { spawn } = require('child_process');

const tsup = spawn('tsup', ['--watch'], { shell: true });

tsup.stdout.on('data', (data) => {
  const output = data.toString();
  process.stdout.write(output);

  if (output.includes('CJS ⚡️ Build success') || output.includes('ESM ⚡️ Build success')) {
    process.stdout.write('\n !!! Build not yet completed! Waiting for DTS !!! \n');
  }

  if (output.includes('DTS ⚡️ Build success')) {
    spawn('pnpm yalc-release', { shell: true, stdio: 'inherit' });
  }
});

tsup.stderr.on('data', (data) => {
  process.stderr.write(data.toString());
});
```
